util/linuxfw/iptables.go

```diff
@@ -823,3 +825,479 @@ type markTarget struct {
+// a local Tailscale IP.
+func (i *iptableRunner) AddLoopbackRule(addr netip.Addr) error {
+	nf := i.ipt4
+	if addr.Is6() {
+		nf = i.ipt6
+	}
+
+	if err := nf.Insert("filter", "ts-input", 1, "-i", "lo", "-s", addr.String(), "-j", "ACCEPT"); err != nil {
+		return fmt.Errorf("adding loopback allow rule for %q: %w", addr, err)
+	}
+
+	return nil
+}
+
+// tsChain returns the name of the tailscale sub-chain corresponding
+// to the given "parent" chain (e.g. INPUT, FORWARD, ...).
+func tsChain(chain string) string {
+	return "ts-" + strings.ToLower(chain)
+}
+
+// getIptablebyProto returns the iptablesInterface in iptables runner for the given protocol.
+func (i *iptableRunner) getIptablebyProto(proto ProtoVersion) iptablesInterface {
+	switch proto {
+	case IPv4:
+		return i.ipt4
+	case IPv6:
+		return i.ipt6
+	default:
+		panic(fmt.Sprintf("unknown protocol %v", proto))
+	}
+}
+
+// DelLoopbackRule removes the iptables rule permitting loopback
+// traffic to a Tailscale IP.
+func (i *iptableRunner) DelLoopbackRule(addr netip.Addr) error {
+	nf := i.ipt4
+	if addr.Is6() {
+		nf = i.ipt6
+	}
+
+	if err := nf.Delete("filter", "ts-input", "-i", "lo", "-s", addr.String(), "-j", "ACCEPT"); err != nil {
+		return fmt.Errorf("deleting loopback allow rule for %q: %w", addr, err)
+	}
+
+	return nil
+}
+
+// GetTables gets the available iptablesInterface in iptables runner.
+func (i *iptableRunner) getTables() []iptablesInterface {
+	if i.GetV6Available() {
+		return []iptablesInterface{i.ipt4, i.ipt6}
+	}
+	return []iptablesInterface{i.ipt4}
+}
+
+// addNetfilterHooks inserts calls to tailscale's netfilter chains in
+// the relevant main netfilter chains. The tailscale chains must
+// already exist.
+func (i *iptableRunner) AddNetfilterHooks() error {
+	divert := func(ipt iptablesInterface, table, chain string) error {
+		tsChain := tsChain(chain)
+
+		args := []string{"-j", tsChain}
+		exists, err := ipt.Exists(table, chain, args...)
+		if err != nil {
+			return fmt.Errorf("checking for %v in %s/%s: %w", args, table, chain, err)
+		}
+		if exists {
+			return nil
+		}
+		if err := ipt.Insert(table, chain, 1, args...); err != nil {
+			return fmt.Errorf("adding %v in %s/%s: %w", args, table, chain, err)
+		}
+		return nil
+	}
+
+	for _, ipt := range i.getTables() {
+		if err := divert(ipt, "filter", "INPUT"); err != nil {
+			return err
+		}
+		if err := divert(ipt, "filter", "FORWARD"); err != nil {
+			return err
+		}
+	}
+	if err := divert(i.ipt4, "nat", "POSTROUTING"); err != nil {
+		return err
+	}
+	if i.v6NATAvailable {
+		if err := divert(i.ipt6, "nat", "POSTROUTING"); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+// AddNetfilterChains creates custom Tailscale chains in netfilter via iptables.
+func (i *iptableRunner) AddNetfilterChains() error {
+	create := func(ipt iptablesInterface, table, chain string) error {
+		err := ipt.ClearChain(table, chain)
+		if errCode(err) == 1 {
```

Comment:
Optional: consider making `errCode(err) == 1` be a function `isErrChainNotExist` or something. But if this is the only place we do this, its probably not worth it.

Reply:
Yeah that would make the code self explanatory, thank for pointing out